backend/src/v2/driver/driver.go

```diff
@@ -2006,7 +2028,253 @@ func resolveUpstreamArtifacts(cfg resolveUpstreamOutputsConfig) (*pipelinespec.A
+		// extract fields from the struct with the wrong key. Hence, the
+		// condition below. NOTE: This is only an issue for Parameter resolution
+		// and does not interfere with Artifact resolution
+		if currentTask.TaskName() == previousTaskName {
+			workingOutputKey = tempWorkingOutputKey
+		}
+
+		previousTaskName = currentTask.TaskName()
+		tempWorkingOutputKey = workingOutputKey
+		workingSubTaskName, workingOutputKey, _ = InspectProducerTask(currentTask, tasks, workingSubTaskName, workingOutputKey, isArtifact)
+
+		glog.V(4).Infof("currentTask ID: %v", currentTask.GetID())
+		glog.V(4).Infof("currentTask Name: %v", currentTask.TaskName())
+		glog.V(4).Infof("currentTask Type: %v", currentTask.GetExecution().GetType())
+		glog.V(4).Infof("workingSubTaskName %v", workingSubTaskName)
+		glog.V(4).Infof("workingOutputKey: %v", workingOutputKey)
+
+		iterations := currentTask.GetExecution().GetCustomProperties()["iteration_count"]
+		iterationIndex := currentTask.GetExecution().GetCustomProperties()["iteration_index"]
+
+		// Base cases for handling the task that actually maps to the task that
+		// created the artifact/parameter we are searching for
+		// Base case 1: currentTask is a ContainerExecution that we can load the values off of
+		if *currentTask.GetExecution().Type == "system.ContainerExecution" {
+			glog.V(4).Infof("currentTask, %v, is a ContainerExecution", currentTaskName)
+			if isArtifact {
+				outputArtifacts, err := cfg.mlmd.GetOutputArtifactsByExecutionId(cfg.ctx, currentTask.GetID())
+				if err != nil {
+					return nil, nil, err
+				}
+				glog.V(4).Infof("outputArtifacts: %#v", outputArtifacts)
+				glog.V(4).Infof("outputKey: %v", workingOutputKey)
+				runtimeArtifact, err := outputArtifacts[workingOutputKey].ToRuntimeArtifact()
+				if err != nil {
+					return nil, nil, cfg.err(err)
+				}
+				glog.V(4).Infof("runtimeArtifact: %v", runtimeArtifact)
+				parallelForArtifactList = append(parallelForArtifactList, runtimeArtifact)
+			} else {
+				_, outputParameters, err := currentTask.GetParameters()
+				glog.V(4).Infof("outputParameters: %v", outputParameters)
+				if err != nil {
+					return nil, nil, cfg.err(err)
+				}
+				parallelForParameterList = append(parallelForParameterList, outputParameters[workingOutputKey])
+			}
+			continue
+
+			// Base case 2: currentTask is a DAGExecution within a loop but is NOT a ParallelFor
+			// Head DAG
+		} else if iterations == nil {
+			tempSubTaskName := workingSubTaskName
+			if iterationIndex != nil {
+				// handle for parallel iteration dag
+				tempSubTaskName = metadata.GetParallelForTaskName(tempSubTaskName, iterationIndex.GetIntValue())
+				glog.V(4).Infof("subTaskIterationName: %v", tempSubTaskName)
+			}
+			glog.V(4).Infof("tempSubTaskName: %v", tempSubTaskName)
+			resolveTaskNameQueue = append(resolveTaskNameQueue, tempSubTaskName)
+			continue
+		}
+
+		for i := range iterations.GetIntValue() {
```

Comment:
Consider adding an explicit conditional for this or a comment that explains that if the if / else if are not met, then we're definitely dealing with a parallel iteration.